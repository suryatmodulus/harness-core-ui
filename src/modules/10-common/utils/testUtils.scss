/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.browser {
  border: 1px solid #cccccc;
  border-radius: 8px;
  min-height: 300px;
  overflow: auto;

  .browser-header {
    background-color: #cccccc;
    padding: 8px 16px;
    display: flex;
  }

  .browser-path {
    display: block;
    border: none;
    line-height: 14px;
    font-size: 12px;
    color: #444444;
    border-radius: 9px;
    min-width: 100px;
    padding: 2px 8px;
    flex: 1 1 auto;
    border: 1px solid transparent;

    &:focus {
      border-color: rgb(30, 167, 253);
      outline: none;
    }
  }

  .browser-content {
    padding: 8px;
  }
}
