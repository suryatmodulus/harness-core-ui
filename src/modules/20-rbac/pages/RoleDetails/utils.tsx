/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
 */

import produce from 'immer'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import type { Permission, PermissionResponse } from 'services/rbac'

export const getPermissionMap = (permissionList?: PermissionResponse[]): Map<ResourceType, Permission[]> => {
  const permissionsMap: Map<ResourceType, Permission[]> = new Map()
  permissionList?.map(response => {
    if (response.permission.resourceType) {
      if (permissionsMap.has(response.permission.resourceType as ResourceType)) {
        permissionsMap.get(response.permission.resourceType as ResourceType)?.push(response.permission)
      } else permissionsMap.set(response.permission.resourceType as ResourceType, [response.permission])
    }
  })
  return permissionsMap
}

export const onPermissionChange = (
  permission: string,
  isAdd: boolean,
  permissions: string[],
  setPermissions: (value: React.SetStateAction<string[]>) => void
): void => {
  if (isAdd) {
    if (
      permission === PermissionIdentifier.EDIT_DASHBOARD ||
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    ) {
      setPermissions(_permissions => [...permissions, permission, PermissionIdentifier.VIEW_DASHBOARD])
    } else {
      setPermissions(_permissions => [...permissions, permission])
    }
  } else if (
    !(
      permission === PermissionIdentifier.VIEW_DASHBOARD &&
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    )
  ) {
    setPermissions(_permissions =>
      produce(_permissions, draft => {
        draft?.splice(permissions.indexOf(permission), 1)
      })
    )
  }
}
