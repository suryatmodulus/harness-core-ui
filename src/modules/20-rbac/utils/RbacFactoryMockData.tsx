/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { ResourceHandler, ResourceCategoryHandler } from '@rbac/factories/RbacFactory'
import { ResourceType, ResourceCategory } from '@rbac/interfaces/ResourceType'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import { String } from 'framework/strings'

export const getResourceTypeHandlerMock = (resource: ResourceType): ResourceHandler | undefined => {
  switch (resource) {
    case ResourceType.PROJECT:
      return {
        icon: 'nav-project',
        label: 'projectsText',
        category: ResourceCategory.ADMINSTRATIVE_FUNCTIONS,
        permissionLabels: {
          [PermissionIdentifier.UPDATE_PROJECT]: <String stringID="rbac.permissionLabels.createEdit" />,
          [PermissionIdentifier.VIEW_PROJECT]: <String stringID="rbac.permissionLabels.view" />,
          [PermissionIdentifier.DELETE_PROJECT]: <String stringID="rbac.permissionLabels.delete" />
        },
        // eslint-disable-next-line react/display-name
        addResourceModalBody: () => <></>
      }
    case ResourceType.ORGANIZATION:
      return {
        icon: 'settings',
        label: 'orgsText',
        category: ResourceCategory.ADMINSTRATIVE_FUNCTIONS,
        permissionLabels: {
          [PermissionIdentifier.UPDATE_ORG]: <String stringID="rbac.permissionLabels.createEdit" />,
          [PermissionIdentifier.VIEW_ORG]: <String stringID="rbac.permissionLabels.view" />,
          [PermissionIdentifier.DELETE_ORG]: <String stringID="rbac.permissionLabels.delete" />
        },
        // eslint-disable-next-line react/display-name
        addResourceModalBody: () => <></>
      }
    case ResourceType.SECRET:
      return {
        icon: 'lock',
        label: 'common.secrets',
        category: ResourceCategory.SHARED_RESOURCES,
        // eslint-disable-next-line react/display-name
        addResourceModalBody: () => <></>
      }
    case ResourceType.DELEGATE:
      return {
        icon: 'main-delegates',
        label: 'delegate.delegates',
        category: ResourceCategory.SHARED_RESOURCES,
        // eslint-disable-next-line react/display-name
        addResourceModalBody: () => <></>
      }
  }
}

export const getResourceGroupTypeHandlerMock = (
  resource: ResourceType | ResourceCategory
): ResourceCategoryHandler | undefined => {
  switch (resource) {
    case ResourceCategory.SHARED_RESOURCES:
      return {
        icon: 'nav-project',
        label: 'rbac.categoryLabels.sharedResources'
      }
    case ResourceCategory.ADMINSTRATIVE_FUNCTIONS:
      return {
        icon: 'nav-project',
        label: 'adminFunctions'
      }
  }
}

export const getResourceCategoryListMock = (): Map<ResourceCategory | ResourceType, ResourceType[] | undefined> => {
  const mockMap = new Map()
  mockMap.set(ResourceCategory.ADMINSTRATIVE_FUNCTIONS, [ResourceType.ORGANIZATION, ResourceType.PROJECT])
  mockMap.set(ResourceCategory.SHARED_RESOURCES, [
    ResourceType.SECRET,
    ResourceType.CONNECTOR,
    ResourceType.DELEGATE,
    ResourceType.DELEGATECONFIGURATION
  ])
  return mockMap
}
