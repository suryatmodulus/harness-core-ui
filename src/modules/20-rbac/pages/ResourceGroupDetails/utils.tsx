import produce from 'immer'
import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import { isDynamicResourceSelector } from '@rbac/utils/utils'
import type { ResourceSelector } from 'services/resourcegroups'

export const getSelectedResourcesMap = (
  resourceSelectorList?: ResourceSelector[]
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()

  resourceSelectorList?.map(resourceSelector => {
    if (isDynamicResourceSelector(resourceSelector.type)) {
      map.set(resourceSelector.resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    } else {
      map.set(resourceSelector.resourceType, resourceSelector.identifiers)
    }
  })
  return map
}

export const validateResourceSelectors = (value: ResourceSelector[]): ResourceType[] => {
  return value
    .filter(val => {
      if (!isDynamicResourceSelector(val.type)) {
        if (val.identifiers?.length < 1) {
          return true
        }
      }
      return false
    })
    .map(val => val.resourceType)
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelector[] => {
  const resourceSelectors: ResourceSelector[] = []
  map.forEach((value, key) => {
    if (isDynamicResourceSelector(value)) {
      resourceSelectors.push({
        type: RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR,
        resourceType: key
      })
    } else {
      resourceSelectors.push({
        type: RbacResourceGroupTypes.STATIC_RESOURCE_SELECTOR,
        resourceType: key,
        identifiers: value
      })
    }
  })
  return resourceSelectors
}

export const computeResourceMapOnChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  resourceType: ResourceType,
  isAdd: boolean,
  identifiers?: string[]
): void => {
  if (identifiers) {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, identifiers)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          if (Array.isArray(draft.get(resourceType))) {
            draft.set(
              resourceType,
              (draft.get(resourceType) as string[]).filter(el => !identifiers.includes(el))
            )
          }
          if (draft.get(resourceType)?.length === 0) {
            draft.delete(resourceType)
          }
        })
      )
    }
  } else {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.delete(resourceType)
        })
      )
    }
  }
}

export const computeResourceMapOnMultiChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  types: ResourceType[],
  isAdd: boolean
): void => {
  if (isAdd) {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      })
    )
  } else {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.delete(resourceType)
        })
      })
    )
  }
}
