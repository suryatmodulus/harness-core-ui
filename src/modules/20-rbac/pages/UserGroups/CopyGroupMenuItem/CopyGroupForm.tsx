/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState } from 'react'
import {
  FormInput,
  Layout,
  Button,
  ButtonVariation,
  Container,
  Formik,
  ModalErrorHandlerBinding,
  FormikForm as Form,
  useToaster,
  ModalErrorHandler,
  MultiSelectOption
} from '@wings-software/uicore'
import { useParams } from 'react-router-dom'
import {
  useGetOrganizationAggregateDTOList,
  useGetProjectListWithMultiOrgFilter,
  useCopyUserGroup
} from 'services/cd-ng'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { useStrings } from 'framework/strings'
import { generateScopeList } from '@rbac/utils/utils'
import type { ProjectSelectOption } from '@audit-trail/components/FilterDrawer/FilterDrawer'
import { getOrgDropdownList, getProjectDropdownList } from '@audit-trail/utils/RequestUtil'
import css from './CopyGroupForm.module.scss'

export interface CopyGroupFormType {
  organizations: MultiSelectOption[]
  projects: ProjectSelectOption[]
}

interface CopyGroupFormProps {
  closeModal: () => void
  identifier: string
}

const CopyGroupForm: React.FC<CopyGroupFormProps> = ({ closeModal, identifier }) => {
  const [orgQuery, setOrgQuery] = useState('')
  const [projectsQuery, setProjectsQuery] = useState('')
  const { accountId } = useParams<ProjectPathProps>()
  const [modalErrorHandler, setModalErrorHandler] = useState<ModalErrorHandlerBinding>()
  const { getString } = useStrings()

  const { data } = useGetOrganizationAggregateDTOList({
    queryParams: {
      pageIndex: 0,
      searchTerm: orgQuery,
      accountIdentifier: accountId
    }
  })

  const { data: projectData, refetch: refetchProjectList } = useGetProjectListWithMultiOrgFilter({
    queryParams: {
      accountIdentifier: accountId,
      searchTerm: projectsQuery,
      orgIdentifiers: []
    },
    lazy: true,
    queryParamStringifyOptions: {
      arrayFormat: 'repeat'
    },
    debounce: 300
  })

  const { mutate: copyUserGroup, loading } = useCopyUserGroup({
    queryParams: { accountIdentifier: accountId, groupIdentifier: identifier }
  })
  const { showSuccess } = useToaster()

  const orgList = data?.data?.content ? getOrgDropdownList(data?.data?.content) : []
  const projects = projectData?.data?.content ? getProjectDropdownList(projectData?.data?.content) : []

  const onSave = async (values: CopyGroupFormType) => {
    modalErrorHandler?.hide()
    const { organizations: selectedOrgs, projects: selectedProjects } = values
    const scopeList = generateScopeList(selectedOrgs, selectedProjects, accountId)
    try {
      const response = await copyUserGroup(scopeList)
      if (response) {
        showSuccess(getString('rbac.copyGroupSuccess'))
      }
      closeModal()
    } catch (e) {
      modalErrorHandler?.showDanger(e.data?.message || e.message)
    }
  }

  return (
    <>
      <Formik<CopyGroupFormType>
        initialValues={{
          organizations: [],
          projects: []
        }}
        formName="copyGroupForm"
        onSubmit={values => {
          onSave(values)
        }}
      >
        {formik => {
          const selectedOrgs = formik.values.organizations

          return (
            <Form>
              <ModalErrorHandler bind={setModalErrorHandler} />
              <Container className={css.copyGroupForm}>
                <FormInput.MultiSelect
                  name="organizations"
                  label={getString('orgLabel')}
                  items={orgList}
                  multiSelectProps={{
                    allowCreatingNewItems: false,
                    onQueryChange: (query: string) => {
                      setOrgQuery(query)
                    }
                  }}
                  onChange={org => {
                    refetchProjectList({
                      queryParams: {
                        accountIdentifier: accountId,
                        orgIdentifiers: org.map(o => o.value as string),
                        pageIndex: 0
                      }
                    })
                  }}
                />
                <FormInput.MultiSelect
                  label={getString('projectLabel')}
                  name="projects"
                  items={projects}
                  disabled={selectedOrgs.length === 0}
                  multiSelectProps={{
                    allowCreatingNewItems: false,
                    onQueryChange: (query: string) => {
                      setProjectsQuery(query)
                    }
                  }}
                />
              </Container>
              <Layout.Horizontal spacing="small">
                <Button
                  disabled={selectedOrgs.length === 0 || loading}
                  type="submit"
                  variation={ButtonVariation.PRIMARY}
                  text={getString('save')}
                />
                <Button variation={ButtonVariation.TERTIARY} onClick={closeModal} text={getString('cancel')} />
              </Layout.Horizontal>
            </Form>
          )
        }}
      </Formik>
    </>
  )
}

export default CopyGroupForm
