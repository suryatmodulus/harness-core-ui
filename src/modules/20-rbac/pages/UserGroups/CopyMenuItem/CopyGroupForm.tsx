/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState } from 'react'
import {
  FormInput,
  Layout,
  Button,
  ButtonVariation,
  Container,
  Formik,
  ModalErrorHandlerBinding,
  FormikForm as Form,
  useToaster,
  ModalErrorHandler
} from '@wings-software/uicore'
import { useParams } from 'react-router-dom'
import {
  useGetOrganizationAggregateDTOList,
  useGetProjectListWithMultiOrgFilter,
  useCopyUserGroup
} from 'services/cd-ng'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import type { CopyGroupFormType } from './CopyMenuItem'
import { useStrings } from 'framework/strings'
import css from './CopyMenuItem.module.scss'
import { getOrgList, getProjectList, getScopeList } from '@rbac/utils/utils'

interface CopyGroupFormProps {
  handleFormClose: () => void
}

const CopyGroupForm: React.FC<CopyGroupFormProps> = ({ handleFormClose }) => {
  const [orgQuery, setOrgQuery] = useState('')
  const [projectsQuery, setProjectsQuery] = useState('')
  const { accountId } = useParams<ProjectPathProps>()
  const [modalErrorHandler, setModalErrorHandler] = useState<ModalErrorHandlerBinding>()
  const { getString } = useStrings()

  const { data } = useGetOrganizationAggregateDTOList({
    queryParams: {
      pageIndex: 0,
      searchTerm: orgQuery,
      accountIdentifier: accountId
    }
  })

  const { data: projectData, refetch: refetchProjectList } = useGetProjectListWithMultiOrgFilter({
    queryParams: {
      accountIdentifier: accountId,
      searchTerm: projectsQuery,
      orgIdentifiers: []
    },
    lazy: true,
    queryParamStringifyOptions: {
      arrayFormat: 'repeat'
    },
    debounce: 300
  })

  const { mutate: copyUserGroup } = useCopyUserGroup({ queryParams: { accountIdentifier: accountId } })
  const { showSuccess } = useToaster()

  const orgList = data?.data?.content ? getOrgList(data?.data?.content) : []
  const projects = projectData?.data?.content ? getProjectList(projectData?.data?.content) : []

  const onSave = async (values: CopyGroupFormType) => {
    modalErrorHandler?.hide()
    const { organizations: selectedOrgs, projects: selectedProjects } = values
    const scopeList = getScopeList(selectedOrgs, selectedProjects, accountId)
    try {
      const response = await copyUserGroup(scopeList)
      if (response) {
        showSuccess(getString('rbac.copyGroupSuccess'))
      }
    } catch (e) {
      modalErrorHandler?.showDanger(e.data?.message || e.message)
    }
  }

  return (
    <>
      <Formik<CopyGroupFormType>
        initialValues={{
          organizations: [],
          projects: []
        }}
        formName="copyGroupForm"
        onSubmit={values => {
          onSave(values)
        }}
      >
        {formik => {
          const selectedOrgs = formik.values.organizations

          return (
            <Form>
              <ModalErrorHandler bind={setModalErrorHandler} />
              <Container className={css.copyGroupForm}>
                <FormInput.MultiSelect
                  name="organizations"
                  label={getString('orgLabel')}
                  items={orgList}
                  multiSelectProps={{
                    allowCreatingNewItems: false,
                    onQueryChange: (query: string) => {
                      setOrgQuery(query)
                    }
                  }}
                  onChange={org => {
                    refetchProjectList({
                      queryParams: {
                        accountIdentifier: accountId,
                        orgIdentifiers: org.map(o => o.value as string),
                        pageIndex: 0
                      }
                    })
                  }}
                />
                <FormInput.MultiSelect
                  label={getString('projectLabel')}
                  name="projects"
                  items={projects}
                  disabled={selectedOrgs.length === 0}
                  multiSelectProps={{
                    allowCreatingNewItems: false,
                    onQueryChange: (query: string) => {
                      setProjectsQuery(query)
                    }
                  }}
                />
              </Container>
              <Layout.Horizontal spacing="small">
                <Button
                  disabled={selectedOrgs.length === 0}
                  type="submit"
                  variation={ButtonVariation.PRIMARY}
                  text={getString('save')}
                />
                <Button variation={ButtonVariation.TERTIARY} onClick={handleFormClose} text={getString('cancel')} />
              </Layout.Horizontal>
            </Form>
          )
        }}
      </Formik>
    </>
  )
}

export default CopyGroupForm
