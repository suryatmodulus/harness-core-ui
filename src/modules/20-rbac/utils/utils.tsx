/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { ReactNode } from 'react'
import { IconName, ModalErrorHandlerBinding, getErrorInfoFromErrorObject, SelectOption } from '@wings-software/uicore'
import { defaultTo } from 'lodash-es'
import type { StringsMap } from 'stringTypes'
import type { AccessControlCheckError, RoleAssignmentMetadataDTO, UserMetadataDTO } from 'services/cd-ng'
import { Scope } from '@common/interfaces/SecretsInterface'
import type {
  Assignment,
  RoleOption,
  ResourceGroupOption
} from '@rbac/modals/RoleAssignmentModal/views/UserRoleAssigment'
import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import RBACTooltip from '@rbac/components/RBACTooltip/RBACTooltip'
import type { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import type { FeatureRequest } from 'framework/featureStore/featureStoreUtil'
import type { PermissionsRequest } from '@rbac/hooks/usePermission'
import { FeatureWarningTooltip } from '@common/components/FeatureWarning/FeatureWarningWithTooltip'
import type { UseStringsReturn } from 'framework/strings'
import css from './utils.module.scss'

export enum PrincipalType {
  USER = 'USER',
  USER_GROUP = 'USER_GROUP',
  SERVICE = 'SERVICE_ACCOUNT'
}

export const getRoleIcon = (roleIdentifier: string): IconName => {
  switch (roleIdentifier) {
    case '_account_viewer':
    case '_organization_viewer':
    case '_project_viewer':
      return 'viewerRole'
    case '_account_admin':
    case '_organization_admin':
    case '_project_admin':
      return 'adminRole'
    case '_pipeline_executor':
      return 'pipeline-executor'
    default:
      return 'customRole'
  }
}

export enum InvitationStatus {
  USER_INVITED_SUCCESSFULLY = 'USER_INVITED_SUCCESSFULLY',
  USER_ADDED_SUCCESSFULLY = 'USER_ADDED_SUCCESSFULLY',
  USER_ALREADY_ADDED = 'USER_ALREADY_ADDED',
  USER_ALREADY_INVITED = 'USER_ALREADY_INVITED',
  FAIL = 'FAIL'
}

interface HandleInvitationResponse {
  responseType: InvitationStatus
  getString: (key: keyof StringsMap, vars?: Record<string, any> | undefined) => string
  showSuccess: (message: string | ReactNode, timeout?: number, key?: string) => void
  modalErrorHandler?: ModalErrorHandlerBinding
  onSubmit?: () => void
  onUserAdded?: () => void
}

export const handleInvitationResponse = ({
  responseType,
  getString,
  showSuccess,
  modalErrorHandler,
  onSubmit,
  onUserAdded
}: HandleInvitationResponse): void => {
  switch (responseType) {
    case InvitationStatus.USER_INVITED_SUCCESSFULLY: {
      onSubmit?.()
      return showSuccess(getString('rbac.usersPage.invitationSuccess'))
    }
    case InvitationStatus.USER_ADDED_SUCCESSFULLY: {
      onUserAdded?.()
      return showSuccess(getString('rbac.usersPage.userAddedSuccess'))
    }
    case InvitationStatus.USER_ALREADY_ADDED:
      return showSuccess(getString('rbac.usersPage.userAlreadyAdded'))
    case InvitationStatus.USER_ALREADY_INVITED:
      return showSuccess(getString('rbac.usersPage.userAlreadyInvited'))
    default:
      return modalErrorHandler?.showDanger(getString('rbac.usersPage.invitationError'))
  }
}

export const getScopeBasedManagedResourceGroup = (
  scope: Scope,
  getString: UseStringsReturn['getString']
): SelectOption => {
  switch (scope) {
    case Scope.ACCOUNT:
      return {
        label: getString('rbac.allAccountResources'),
        value: '_all_account_level_resources'
      }
    case Scope.ORG:
      return {
        label: getString('rbac.allOrgResources'),
        value: '_all_organization_level_resources'
      }
    case Scope.PROJECT:
      return {
        label: getString('rbac.allProjectResources'),
        value: '_all_project_level_resources'
      }
    default:
      return {
        label: getString('rbac.allResources'),
        value: '_all_resources'
      }
  }
}

export const getScopeBasedDefaultAssignment = (
  scope: Scope,
  getString: UseStringsReturn['getString'],
  isCommunity: boolean
): Assignment[] => {
  if (isCommunity) {
    return []
  } else {
    const resourceGroup: ResourceGroupOption = {
      managedRoleAssignment: true,
      ...getScopeBasedManagedResourceGroup(scope, getString)
    }
    switch (scope) {
      case Scope.ACCOUNT:
        return [
          {
            role: {
              label: getString('common.accViewer'),
              value: '_account_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      case Scope.ORG:
        return [
          {
            role: {
              label: getString('common.orgViewer'),
              value: '_organization_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      case Scope.PROJECT:
        return [
          {
            role: {
              label: getString('common.projectViewer'),
              value: '_project_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      default:
        return []
    }
  }
}

export const isAssignmentFieldDisabled = (value: RoleOption | ResourceGroupOption): boolean => {
  if (value.assignmentIdentifier || value.managedRoleAssignment) {
    return true
  }
  return false
}
export const isDynamicResourceSelector = (value: string | string[]): boolean => {
  if (value === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR) {
    return true
  }
  return false
}

interface ErrorHandlerProps {
  data: AccessControlCheckError
}

export const getRBACErrorMessage = (error: ErrorHandlerProps): string | React.ReactElement => {
  const err = error?.data
  if (err?.code === 'NG_ACCESS_DENIED' && err?.failedPermissionChecks?.length) {
    const { permission, resourceType, resourceScope } = err.failedPermissionChecks[0]
    if (permission && resourceType && resourceScope) {
      return (
        <RBACTooltip
          permission={permission as PermissionIdentifier}
          resourceType={resourceType as ResourceType}
          resourceScope={resourceScope}
          className={css.tooltip}
        />
      )
    }
  }
  return getErrorInfoFromErrorObject(error)
}

export const getAssignments = (roleBindings: RoleAssignmentMetadataDTO[]): Assignment[] => {
  return (
    roleBindings?.map(roleAssignment => {
      return {
        role: {
          label: roleAssignment.roleName,
          value: roleAssignment.roleIdentifier,
          managed: roleAssignment.managedRole,
          assignmentIdentifier: roleAssignment.identifier,
          managedRoleAssignment: roleAssignment.managedRoleAssignment
        },
        resourceGroup: {
          label: roleAssignment.resourceGroupName || '',
          value: roleAssignment.resourceGroupIdentifier || '',
          managedRoleAssignment: roleAssignment.managedRoleAssignment,
          assignmentIdentifier: roleAssignment.identifier
        }
      }
    }) || []
  )
}

export const isNewRoleAssignment = (assignment: Assignment): boolean => {
  return !(assignment.role.assignmentIdentifier || assignment.resourceGroup.assignmentIdentifier)
}

interface FeatureProps {
  featureRequest: FeatureRequest
  isPermissionPrioritized?: boolean
}

interface TooltipProps {
  permissionRequest?: Omit<PermissionsRequest, 'permissions'> & { permission: PermissionIdentifier }
  featureProps?: FeatureProps
  canDoAction: boolean
  featureEnabled: boolean
}

interface TooltipReturn {
  tooltip?: React.ReactElement
}

export function getTooltip({
  permissionRequest,
  featureProps,
  canDoAction,
  featureEnabled
}: TooltipProps): TooltipReturn {
  // if permission check override the priorirty
  if (featureProps?.isPermissionPrioritized && permissionRequest && !canDoAction) {
    return {
      tooltip: (
        <RBACTooltip
          permission={permissionRequest.permission}
          resourceType={permissionRequest.resource.resourceType}
          resourceScope={permissionRequest.resourceScope}
        />
      )
    }
  }

  // feature check by default take priority
  if (featureProps?.featureRequest && !featureEnabled) {
    return {
      tooltip: <FeatureWarningTooltip featureName={featureProps?.featureRequest.featureName} />
    }
  }

  // permission check
  if (permissionRequest && !canDoAction) {
    return {
      tooltip: (
        <RBACTooltip
          permission={permissionRequest.permission}
          resourceType={permissionRequest.resource.resourceType}
          resourceScope={permissionRequest.resourceScope}
        />
      )
    }
  }

  return {}
}

export const getUserName = (user: UserMetadataDTO): string => {
  return defaultTo(user.name, user.email)
}
