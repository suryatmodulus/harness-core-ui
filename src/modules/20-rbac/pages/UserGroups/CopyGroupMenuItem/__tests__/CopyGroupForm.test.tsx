/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { TestWrapper } from '@common/utils/testUtils'
import CopyGroupForm from '../CopyGroupForm'

describe('Copy group form test', () => {
  test('render', () => {
    const renderObj = (
      <TestWrapper>
        <CopyGroupForm closeModal={jest.fn} identifier="dummy_identifier" />
      </TestWrapper>
    )
    expect(renderObj).toMatchSnapshot()
  })
})
