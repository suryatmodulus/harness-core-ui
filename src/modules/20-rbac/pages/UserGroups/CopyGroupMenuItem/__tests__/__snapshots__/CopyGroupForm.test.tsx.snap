// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copy group form test render 1`] = `
<div>
  <div
    class="OverlaySpinner--overlaySpinner"
  >
    <form
      class="FormikForm--main"
    >
      <div>
        <div
          class="StyledProps--font StyledProps--main copyGroupForm"
        >
          <div
            class="bp3-form-group"
            data-id="organizations-0"
          >
            <label
              class="bp3-label"
              for="organizations"
            >
              <span
                class="Tooltip--acenter"
                data-tooltip-id="copyGroupForm_organizations"
              >
                orgLabel
              </span>
               
              <span
                class="bp3-text-muted"
              />
            </label>
            <div
              class="bp3-form-content"
            >
              <div
                class="bp3-popover-wrapper MultiSelect--main bp3-fill"
              >
                <div
                  class="bp3-popover-target"
                >
                  <div
                    class=""
                  >
                    <div
                      class="bp3-input bp3-tag-input bp3-multi-select"
                    >
                      <div
                        class="bp3-tag-input-values"
                      >
                        <input
                          autocomplete="off"
                          class="bp3-input-ghost bp3-multi-select-tag-input-input"
                          name="organizations"
                          placeholder="- Select -"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="bp3-form-group bp3-disabled"
            data-id="projects-1"
          >
            <label
              class="bp3-label"
              for="projects"
            >
              <span
                class="Tooltip--acenter"
                data-tooltip-id="copyGroupForm_projects"
              >
                projectLabel
              </span>
               
              <span
                class="bp3-text-muted"
              />
            </label>
            <div
              class="bp3-form-content"
            >
              <div
                class="bp3-popover-wrapper MultiSelect--main bp3-fill"
              >
                <div
                  class="bp3-popover-target"
                >
                  <div
                    class=""
                  >
                    <div
                      class="bp3-input bp3-tag-input bp3-disabled bp3-multi-select"
                    >
                      <div
                        class="bp3-tag-input-values"
                      >
                        <input
                          autocomplete="off"
                          class="bp3-input-ghost bp3-multi-select-tag-input-input"
                          disabled=""
                          name="projects"
                          placeholder="- Select -"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main"
        >
          <button
            class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-primary"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <span
              class="bp3-button-text"
            >
              save
            </span>
          </button>
          <button
            class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-tertiary"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              cancel
            </span>
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
